### PR TITLE
Metropolis state lazy using State (StdGen, Double)

### PR DESCRIPTION
**Pull request by froth**
I have a kind of working draft without Monad Transformers.

Two things are unfinished:
 - I removed configurable options for stepsize and logprob as I needed to make the interfaces easier for me to understand. But I think this can be reintroduced easily. Also I degeneralized the signatures to Double for the same reason of not seeing the wood for the trees.
 - The new version gives a somewhat different result compared to the Monad Transformer one using the same gen. I am unsure if this is a result of the single State and the generators used in a different way now or if this is a subtle bug that I do not understand. Maybe you can have a close look if something looks implausible. I am not very sure what to expect from the mcmc as I never worked with it :)

Have a nice weekend.

### CODE REVIEW

**Comment by tomatitito** [comment]
Very nice!
- It's interesting that you get different results compared to my StateT solution. I also get different results between the two versions I had (the StateT one and the one using recursion with a stopping criterion). So altogether we have three different results now...
- My solution differs in that only the `transition'` function lives in `State (StdGen, Double) Double`, whereas my `propose'` and `acceptOrReject'` functions are not monadic anymore. I do like your solution better, although I don't fully understand it, yet!

I will merge once I understand it, putting the stuff back in that you removed!

File: src/Kernels/MetropolisHastingsLazy.hs
Comment (tomatitito) [inline src/Kernels/MetropolisHastingsLazy.hs]: @froth Is there a reason for not using `return nextValue` here?

Comment (froth) [reply src/Kernels/MetropolisHastingsLazy.hs]: @tomatitito no reason. `return nextValue` is better.

Comment (tomatitito) [reply src/Kernels/MetropolisHastingsLazy.hs]: @froth Ok cool, that's what threw me off! Also, I think we should `return proposal` while still putting `(newGen, nextValue)`. It's important that `nextValue` is in the state, so it can be retrieved when building up the markov chain. However, `proposal` is the value that is computed, even though it might not be used. I think it doesn't matter for the result (haven't tried yet, though), but it better reflects the algorithm.

Comment (froth) [reply src/Kernels/MetropolisHastingsLazy.hs]: I think the result is ignored by replicateM, yes. That would mean only the last value returned would change. But I think we are only interested in the chain anyways, right?

Comment (froth) [reply src/Kernels/MetropolisHastingsLazy.hs]: I think we could also do return () without changing much?

Comment (tomatitito) [reply src/Kernels/MetropolisHastingsLazy.hs]: Yes, we are only interested in the chain. So you are right, we could just discard the computed value altogether and return (). But I think that's what `put` does for us: setting the state and returning `()`.

Comment (froth) [reply src/Kernels/MetropolisHastingsLazy.hs]: ah nice, learned something new today :)

Comment (froth) [reply src/Kernels/MetropolisHastingsLazy.hs]: @tomatitito  It does not work as expected. If I return () (or put) the signature of transition' becomes `transition' :: State (StdGen, Double) ()` and thereby the signature sampleStateT' becomes `Double -> StdGen -> Int -> ([()], (StdGen, Double))` and by that we loose our chain. Therefore I also think that it is indeed important to `return nextValue`. Maybe this is the difference between the implementations?

File: src/Kernels/MetropolisHastingsLazy.hs
Comment (tomatitito) [inline src/Kernels/MetropolisHastingsLazy.hs]: Let's put the stepsize back in as an argument. The stepsize can be used to fine tune the algorithm to get better results, so it's good to not hardcode it.

File: src/Kernels/MetropolisHastingsLazy.hs
Comment (tomatitito) [inline src/Kernels/MetropolisHastingsLazy.hs]: Let's put the logProb back into the function. In our case, it's `Double -> Double`. The computation for `nextValue` needs to make use of this function. It should be `if rand < acceptanceProbability current proposal logProb`.

File: src/Kernels/MetropolisHastingsLazy.hs
Comment (tomatitito) [inline src/Kernels/MetropolisHastingsLazy.hs]: Let's go back to the old version and have logProb as an argument. This is the function that gives the answer to the question: what is the probability of x? But on the log scale. We can just make it `Double -> Double` for now. The logProb function differs depending on the probability distribution used. That's why hardcoding it is not a good idea. We need to be able to pass the right function here.

File: src/Kernels/MetropolisHastingsLazy.hs
Comment (tomatitito) [inline src/Kernels/MetropolisHastingsLazy.hs]: After having changed `propose'` and `acceptOrReject'`, we need to partially apply them here to make the composition work.

**Review by tomatitito — CHANGES_REQUESTED** [review]
Nice, I like it a lot! I requested some changes, mainly to get the stuff back in that you removed, before merging.

File: src/Kernels/MetropolisHastingsLazy.hs
Comment (froth) [reply src/Kernels/MetropolisHastingsLazy.hs]: done

File: src/Kernels/MetropolisHastingsLazy.hs
Comment (froth) [reply src/Kernels/MetropolisHastingsLazy.hs]: done

File: src/Kernels/MetropolisHastingsLazy.hs
Comment (froth) [reply src/Kernels/MetropolisHastingsLazy.hs]: done

File: src/Kernels/MetropolisHastingsLazy.hs
Comment (froth) [reply src/Kernels/MetropolisHastingsLazy.hs]: done

**Comment by froth** [comment]
I reintroduced stepsize and logProb. I also wrapped them into newtypes to make the interfaces clearer so that it is easy to differentiate between the Double values used in the chain and the Double value of stepsize.

**Comment by tomatitito** [comment]
Very nice! Great idea to introduce newtypes, thanks!

**Pull request merged**
